### ISSUE TITLE
`torch.autograd` prematurely returns zero for finite float64 sigmoid and tanh tail derivatives

### ISSUE THREAD

**Issue by beanduan22** [pytorch/pytorch#195527]
## Describe the bug

`torch.autograd` returns zero first- and second-order derivatives for finite
`float64` inputs in the positive tails of `torch.sigmoid` and `torch.tanh`,
even though the analytic derivatives are finite, normal `float64` numbers.

For sigmoid,

```text
d/dx sigmoid(x) = exp(-x) / (1 + exp(-x))**2
```

so at `x = 40.0` the derivative is approximately
`4.248354255291589e-18`.  PyTorch returns `0.0`.  The corresponding value at
`x = -40.0` remains nonzero, which rules out underflow.

For tanh,

```text
d/dx tanh(x) = 4 * exp(-2x) / (1 + exp(-2x))**2
```

so at `x = 20.0` the derivative is approximately
`1.6993417021166355e-17`.  PyTorch again returns `0.0`.

The behavior is consistent with reconstructing the derivative from the rounded
forward output (`y * (1 - y)` for sigmoid, or `1 - y**2` for tanh).  At these
inputs the forward results round to exactly `1.0`, causing cancellation before
the derivative itself underflows.

This reproduces on CPU.  CUDA availability does not affect the CPU result.

```python
import math
import torch

torch.set_default_dtype(torch.float64)


def derivatives(fn, x_value):
    x = torch.tensor(x_value, requires_grad=True)
    y = fn(x)
    first, = torch.autograd.grad(y, x, create_graph=True)
    second, = torch.autograd.grad(first, x)
    return y.item(), first.item(), second.item()


for name, fn, x_value, expected in (
    (
        "sigmoid",
        torch.sigmoid,
        40.0,
        math.exp(-40.0) / (1.0 + math.exp(-40.0)) ** 2,
    ),
    (
        "tanh",
        torch.tanh,
        20.0,
        4.0 * math.exp(-40.0) / (1.0 + math.exp(-40.0)) ** 2,
    ),
):
    value, first, second = derivatives(fn, x_value)
    print(name)
    print("forward:", value)
    print("first derivative:", first)
    print("second derivative:", second)
    print("expected first derivative:", expected)
```

## Actual result

On PyTorch `2.11.0+cu128` with CPU execution:

```text
sigmoid
forward: 1.0
first derivative: 0.0
second derivative: -0.0
expected first derivative: 4.248354255291589e-18

tanh
forward: 1.0
first derivative: 0.0
second derivative: -0.0
expected first derivative: 1.6993417021166355e-17
```

## Expected result

The derivatives should remain finite and nonzero while their analytic values
are representable.  In particular, the first derivatives above should be close
to `4.248354255291589e-18` and `1.6993417021166355e-17`, respectively.

### Versions

```text
PyTorch version: 2.11.0+cu128
Python version: 3.13.5
Default dtype: float64
Execution device: CPU
```

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93
